### PR TITLE
feat(array): Implement Find and FindAll

### PR DESCRIPTION
**Pull request by k4n4ry**
close https://github.com/gavv/httpexpect/issues/168

### CODE REVIEW

File: object_test.go [hunk 1]
```diff
@@ -1006,3 +1018,267 @@ func TestObjectFilter(t *testing.T) {
...
+		foundValues := object.FindAll(func(key string, value *Value) bool {
+			return key == "baz" || key == "quux"
+		})
+
+		actual := []interface{}{}
+		for _, value := range foundValues {
+			actual = append(actual, value.Raw())
+		}
+
+		assert.ElementsMatch(t, []interface{}{"qux", "corge"}, actual)
```
Comment (k4n4ry) [inline object_test.go:1123]: `map[string]interface{}` has randomize iteration order, so I use `assert.ElementsMatch`

File: array.go [hunk 2]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
...
+
+	a.chain.fail(AssertionFailure{
+		Type:   AssertValid,
+		Actual: &AssertionValue{a.value},
+		Errors: []error{
+			errors.New("expected: value is match in array"),
+		},
+	})
+
+	return newValue(a.chain, []interface{}{})
```
[the remark this answers is not in the document]
Comment (gavv) [reply array.go:412]: Same in a few other places.

File: array.go [hunk 3]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
...
+	if len(foundValues) == 0 {
+		a.chain.fail(AssertionFailure{
+			Type:   AssertValid,
+			Actual: &AssertionValue{a.value},
+			Errors: []error{
+				errors.New("expected: value is match in array"),
+			},
+		})
+		return []Value{}
+	}
```
[the remark this answers is not in the document]
Comment (gavv) [reply array.go:468]: Same for Object.

File: array.go [hunk 4]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
...
+		})
+		valueChain.replace("NotFind[%v]", index)
+		if fn(index, newValue(valueChain, element)) && !chainFailed {
+			a.chain.fail(AssertionFailure{
+				Type:   AssertValid,
+				Actual: &AssertionValue{a.value},
+				Errors: []error{
+					errors.New("expected: value is not match in array"),
+				},
+			})
```
[the remark this answers is not in the document]
Comment (gavv) [reply array.go:516]: Same for Object.

**Review by gavv — COMMENTED** [review]
Thank you! A few comments.

File: object_test.go (hunk 1 above)
Comment (gavv) [reply object_test.go:1123]: Thanks for this notice!

In this lib I try to make all methods deterministic. Object.Keys and Object.Values return elements sorted by key. Let's use the same approach here: sort values by key before looking for them, in all three methods. This way results and error messages will be the same each run.

**Comment by gavv** [comment]
It would be nice to cover two additional cases in unit tests:

* when predicate returns true, but assertion fails, and predicate is considered failed
* for FindAll: same, but when assertion fails only for some of the values (to ensure that assertion failure does not affect subsequent matches)

File: object_test.go (hunk 1 above)
Comment (gavv) [reply object_test.go:1123]: I've also created related issue https://github.com/gavv/httpexpect/issues/212

**Comment by gavv** [comment]
@k4n4ry BTW, welcome to discord chat https://discord.gg/5SCPCuCWA9 :)

**Comment by gavv** [comment]
BTW, coveralls reported reduced coverage.

Could you cover missing branches please? They are new AssertUsage failures.

https://coveralls.io/builds/55447288/source?filename=object.go#L368
https://coveralls.io/builds/55447288/source?filename=array.go#L382

File: object_test.go (hunk 1 above)
Comment (k4n4ry) [reply object_test.go]: fixed [48d4b10](https://github.com/gavv/httpexpect/pull/211/commits/48d4b1018a8acfff08613fb8f35198489df2719f) 
sort values by key before looking for them.

File: array.go [hunk 5]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
 	return newArray(a.chain, array)
 }
 
+// Find returns first matched element,
+// or fails and returns empty (non-nil) value if element was not found
+//
```
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: fixed 48d4b10 .
Sorry if my English is not correct.

File: array.go [hunk 6]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
...
+// Example:
+//
+//	array := NewArray(t, []interface{}{1, "foo", 2, "bar"})
+//	foundValue := array.Find(func(index int, value *httpexpect.Value) bool {
+//		return value.String().Raw() != ""
+//	})
+//	foundValue.Equal("foo")	// succeeds
```
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: fixed 48d4b10 .

File: array.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: I added [679a177](https://github.com/gavv/httpexpect/pull/211/commits/679a1772123ad22aee7e4f1de5b629ee2b6c7d26)

File: array.go [hunk 7]
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
...
+		if fn(index, newValue(valueChain, element)) && !chainFailed {
+			return newValue(a.chain, element)
+		}
+	}
+
+	a.chain.fail(AssertionFailure{
+		Type:   AssertValid,
+		Actual: &AssertionValue{a.value},
+		Errors: []error{
+			errors.New("expected: value is match in array"),
```
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

File: array.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6
I also fixed test case e.g. 
https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6#diff-64f424394477d086503ab2306249df2f723d6ac2b466ba1d1c00f9624560b4a1R996  
https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6#diff-64f424394477d086503ab2306249df2f723d6ac2b466ba1d1c00f9624560b4a1R1209

File: array.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: removed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

File: array.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (k4n4ry) [reply array.go]: I fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

**Comment by k4n4ry** [comment]
thanks for review, and I added test cases for object/array .

> It would be nice to cover two additional cases in unit tests:
> 
> when predicate returns true, but assertion fails, and predicate is considered failed
> for FindAll: same, but when assertion fails only for some of the values (to ensure that assertion failure does not affect subsequent matches)

and added test cases for missing branches.

> Could you cover missing branches please? They are new AssertUsage failures.
> 
> https://coveralls.io/builds/55447288/source?filename=object.go#L368
> https://coveralls.io/builds/55447288/source?filename=array.go#L382

**Comment by gavv** [comment]
Thank you!

Here are two small follow-up commits:

* 8300a097aed00212f121717756a512ae1076de79
* 18da34490c9503020b3bee68dad52647d401062b

See commit messages for the list of changes. They're mostly cosmetic.

**Comment by gavv** [comment]
One more small commit: 0978c258998559dc00df2c87cf186483169778d0

**Pull request merged**